sources/zafira-services/src/main/java/com/qaprosoft/zafira/service/WidgetService.java

```diff
@@ -79,15 +94,95 @@ public Widget getWidgetById(long id) {
+            additionalParams.put(WidgetService.DefaultParam.CURRENT_USER_ID, userId);
+            resultList = executeSQL(widgetTemplate.getSql(), params, additionalParams, true);
+        } catch (Exception e) {
+            if (stackTraceRequired) {
+                resultList = List.of(Map.of("Check your query", ExceptionUtils.getFullStackTrace(e)));
+            } else {
+                // wrap whatever error is thrown
+                throw new ProcessingException(WIDGET_QUERY_EXECUTION_ERROR, e.getMessage(), e);
+            }
+        }
+        return resultList;
+    }
+
+    /**
+        Used for old widget versions and table widgets.
+        Someday we'll remove echarts library completely and refactor this method.
+     */
+
+    @Transactional(readOnly = true)
+    public List<Map<String, Object>> getQueryResultObsolete(List<String> projects,
+                                                            String currentUserId,
+                                                            String dashboardName,
+                                                            boolean stackTraceRequired,
+                                                            String query,
+                                                            List<Attribute> attributes,
+                                                            Long userId,
+                                                            String userName) {
+        List<Map<String, Object>> resultList;
+        try {
+            query = applyAttributes(attributes, query);
+            query = query
+                    .replaceAll("#\\{project}", formatProjects(projects))
+                    .replaceAll("#\\{dashboardName}", !StringUtils.isEmpty(dashboardName) ? dashboardName : "")
+                    .replaceAll("#\\{currentUserId}", !StringUtils.isEmpty(currentUserId) ? currentUserId : String.valueOf(userId))
+                    .replaceAll("#\\{currentUserName}", String.valueOf(userName))
+                    .replaceAll("#\\{zafiraURL}", urlResolver.buildWebURL())
+                    .replaceAll("#\\{hashcode}", "0")
+                    .replaceAll("#\\{testCaseId}", "0");
+
+            resultList = executeSQL(query);
+        } catch (Exception e) {
+            if (stackTraceRequired) {
+                resultList = List.of(Map.of("Check your query", ExceptionUtils.getFullStackTrace(e)));
+            } else {
+                throw new ProcessingException(WIDGET_QUERY_EXECUTION_ERROR, e.getMessage(), e);
+            }
+        }
+        return resultList;
+    }
+
+    private String applyAttributes(List<Attribute> attributes, String query) {
+        if (attributes != null) {
+            for (Attribute attribute : attributes) {
+                query = query.replaceAll("#\\{" + attribute.getKey() + "\\}", attribute.getValue());
+            }
+        }
+        return query;
+    }
+
+    private String formatProjects(List<String> projects) {
```

Comment:
Let's come up with a better name for this one. We are not really formatting projects, but we construct a single string containing project names, so let's come up with appropriate name.